packages/shylu/shylu_dd/frosch/src/SchwarzOperators/FROSch_MultiplicativeOperator_def.hpp

```diff
@@ -0,0 +1,246 @@
+
+namespace FROSch {
+    
+    template <class SC,class LO,class GO,class NO>
+    MultiplicativeOperator<SC,LO,GO,NO>::MultiplicativeOperator(CrsMatrixPtr k, ParameterListPtr parameterList) :
+    SchwarzOperator<SC,LO,GO,NO> (k, parameterList),
+    OperatorVector_ (0),
+    EnableOperators_ (0)
+    {
+        
+    }
+    
+    template <class SC,class LO,class GO,class NO>
+    MultiplicativeOperator<SC,LO,GO,NO>::MultiplicativeOperator(CrsMatrixPtr k, SchwarzOperatorPtrVecPtr operators, ParameterListPtr parameterList) :
+    SchwarzOperator<SC,LO,GO,NO> (k, parameterList),
+    OperatorVector_ (0),
+    EnableOperators_ (0)
+    {
+        OperatorVector_.push_back(operators.at(0));
+        for (unsigned i=1; i<operators.size(); i++) {
+            FROSCH_ASSERT(operators[i]->OperatorDomainMap().SameAs(OperatorVector_[i]->OperatorDomainMap()),"The DomainMaps of the operators are not identical.");
+            FROSCH_ASSERT(operators[i]->OperatorRangeMap().SameAs(OperatorVector_[i]->OperatorRangeMap()),"The RangeMaps of the operators are not identical.");
+            
+            OperatorVector_.push_back(operators[i]);
+            EnableOperators_.push_back(true);
+        }
+    }
+        
+    template <class SC,class LO,class GO,class NO>
+    MultiplicativeOperator<SC,LO,GO,NO>::~MultiplicativeOperator()
+    {
+        
+    }
+    
+    template <class SC,class LO,class GO,class NO>
+    void MultiplicativeOperator<SC,LO,GO,NO>::preApplyCoarse(MultiVector &x, MultiVector &y)
+    {
+        FROSCH_ASSERT(this->OperatorVector_.size()==2,"Should be a Two-Level Operator.");
+        this->OperatorVector_[1]->apply(x,y,true);
+        
+    }
+    
+    // Y = alpha * A^mode * X + beta * Y
+    template <class SC,class LO,class GO,class NO>
+    void MultiplicativeOperator<SC,LO,GO,NO>::apply(const MultiVector &x,
+                                         MultiVector &y,
+                                         bool usePreconditionerOnly,
+                                         Teuchos::ETransp mode,
+                                         SC alpha,
+                                         SC beta) const
+    {
+
+        FROSCH_ASSERT(usePreconditionerOnly,"MultiplicativeOperator can only be used as a preconditioner.");
+        FROSCH_ASSERT(this->OperatorVector_.size()==2,"Should be a Two-Level Operator.");
+
+
+        MultiVectorPtr xTmp = Xpetra::MultiVectorFactory<SC,LO,GO,NO>::Build(x.getMap(),x.getNumVectors());
+        *xTmp = x; // Need this for the case when x aliases y
+
+        MultiVectorPtr yTmp = Xpetra::MultiVectorFactory<SC,LO,GO,NO>::Build(y.getMap(),y.getNumVectors());
+        *yTmp = y; // for the second apply
+
+        this->OperatorVector_[0]->apply(*xTmp,*yTmp,true);
+        
+        this->K_->apply(*yTmp,*xTmp);
+        
+        this->OperatorVector_[1]->apply(*xTmp,*xTmp,true);
+
+        yTmp->update(1.0,*xTmp,-1.0);
+        y.update(alpha,*yTmp,beta);
+        
+        
+    }
```

Comment:
Does the multiplicative operator only work for the case of two operators?

Reply:
Atm that is the case (see assert above). Added to ToDo-List to have a more general operator.